File: LeetCode problem set/14_Longest_Common_Prefix.js
// https://leetcode.com/problems/longest-common-prefix/

const longestCommonPrefix = (arr = []) => {
  let ans = "";
  let n = arr.length;
  if (n == 0) return ans;
  arr.sort((a, b) => {
    if (a > b) return 1;
    else if (a < b) return -1;
    else return 1;
  });

  let first = arr[0];
  let last = arr[n - 1];
  let len = Math.min(first.length, last.length);
  for (let i = 0; i < len; i++) {
    if (first[i] != last[i]) {
      return ans;
    }
    ans += first[i];
  }
  return ans;
};

console.log(longestCommonPrefix(["abab", "aba", "abc"]));

File: LeetCode problem set/14_Longest_Common_Prefix.cpp
// https://leetcode.com/problems/longest-common-prefix/
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  string longestCommonPrefix(vector<string>& strs) {
    string ans = "";
    sort(strs.begin(), strs.end());
    int n = strs.size();
    if ( n == 0) return ans;
    int firstLen = strs[0].size(), lastLen = strs[n - 1].size();
    for (int i = 0; i < min(firstLen, lastLen); i++) {
      if(strs[0][i] != strs[n-1][i]) {
        return ans;
      }
      ans += strs[0][i];
    }
    return ans;
  }
};

int main() {
  Solution sol;
  vector<string> strs = {"flower","flow","flight"};
  cout << sol.longestCommonPrefix(strs) << endl;
  return 0;
}